`--cover assume`: Add assert statements before assume to check for coverage

NlightNFotis opened the pull request:
This PR describes a proposed fix to #6057.

It is a work in progress, as it still needs some extra test cases,
and documentation (and whatever fixes may spring up as part of
a CI run).

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

NlightNFotis committed:
Enfold `assume` in an `assert(false)` to allow for easier debugging.

This allows for easier debugging by allowing `assume(false)`s that
end up emptying the search state space to be identified on a more
granular basis.

NlightNFotis committed:
Add reference test case for `--cover assume` code.

This was provided as reference for an issue described in
https://github.com/diffblue/cbmc/issues/6057.

TGWDB commented on src/goto-checker/properties.h:
```diff
@@ -73,7 +73,7 @@ struct property_infot
 
 /// A map of property IDs to property infos
-typedef std::unordered_map<irep_idt, property_infot> propertiest;
+typedef std::map<irep_idt, property_infot> propertiest;
```
I know there are some performance concerns related to ordered maps. I know this choice was made to "fix" the output order, but maybe this should be checked more carefully since other output does not have this problem?

NlightNFotis replied:
I have consulted with @peterschrammel offline and this doesn't appear to be a problem.

TGWDB commented on src/goto-checker/cover_goals_report_util.cpp:
```diff
@@ -53,9 +62,7 @@ static void output_goals_plain(const propertiest &properties, messaget &log)
-                 << (property_pair.second.status == property_statust::FAIL
-                       ? "SATISFIED"
-                       : "FAILED")
+                 << mapOfResults[property_pair.second.status]
```
This is a subtle change to the outcomes when we see statuses other than `FAIL`. Do we know for sure there are no other parts of the code or output that might rely upon this?

NlightNFotis replied:
Good catch - I have reverted the changes to this file.

TGWDB reviewed (changes requested):
Some comments in the code, two other things (apart from passing existing checks):

1. It would be good to have more examples, also including examples where there are no vacuous paths.
2. doxygen!

NlightNFotis committed:
Add second test case that features assume statement hidden behind
branch.

NlightNFotis committed:
[DOCS] Add documentation for the new `--cover assume` coverage criterion.

TGWDB commented on doc/cprover-manual/modeling-assumptions.md:
```diff
@@ -71,3 +71,37 @@ int main() {
+
+You can ask CBMC to give coverage information regarding `__CPROVER_assume` statements.
+This is useful when you need, for example, to check which assume statements may have
+led to an emptying of the search state space, resulting in `assert` statements being
+vaccuously passing.
```
"being vacuously passing" -> "being vacuously passed".

TGWDB commented on doc/cprover-manual/test-suite.md:
```diff
@@ -213,6 +213,7 @@ The table below summarizes the coverage criteria that CBMC supports.
 Criterion |Definition
 ----------|----------
 assertion |For every assertion, generate a test that reaches it
+assume    |For every assume, generate a test to make sure that the state space before/after it hasn't been emptied
```
Propose:
"For every assume, generate tests before and after the assume statement to indicate coverage before and after the assume statement"

We could also rewrite to be clearer about empty state space, but this is maybe not the clean mechanics, instead focusing on the motivation/theory.

TGWDB commented on src/goto-instrument/cover_instrument_assume.cpp:
```diff
@@ -0,0 +1,52 @@
+/// \file cover_instrument_assume.cpp
+/// Coverage Instrumentation for ASSUME instructions.
```
Should this file have a Diffblue authorship statement?

peterschrammel commented on src/goto-instrument/cover_instrument_assume.cpp:
```diff
@@ -0,0 +1,52 @@
+    const auto location = i_it->source_location;
+    const auto comment = std::string(location.get_comment().c_str());
+    // Filter based on if assertion was added by us as part of instrumentation.
+    if(comment.find("assert(false) ") == std::string::npos)
```
🚫 use `initialize_source_location()` which sets `source_location.set_property_class` to `coverage` and check that.

NlightNFotis replied:
Done 👍🏻

TGWDB committed:
Fix Windows error by changing scope of namespace

This moves a dummy namespace needed for expr2c to be created as
the argument to the call.
